odb_extractor_atoms.py

```diff
@@ -59,98 +223,297 @@ def select_obsclass(classes):
-        fpu = step[fpuinst[inst]]
-        if 'instrument:filter' in step.keys():
-            filter = step["instrument:filter"]
-        else:
-            filter = 'None'
-        wavelength = float(step['instrument:observingWavelength'])
         step_time = step['totalTime'] / 1000.
 
-        if 'GMOS' in inst:
-            coadds = 1
-        else:
-            coadds = int(step['observe:coadds'])
-
-        disperser = step['instrument:disperser']
-        if 'telescope:p' in step.keys():
-            p = float(step['telescope:p'])
-        else:
-            p = 0.0
-        if 'telescope:q' in step.keys():
-            q = float(step['telescope:q'])
-            qoffsets = True
-        else:
-            q = 0.0
-
-            # Any wavelength/filter change is a new atom
-        if ii == 0 or (ii > 0 and wavelength != float(sequence[ii - 1]['instrument:observingWavelength'])):
+        atomstr = 'Atom for: '
+        # Any wavelength/filter change is a new atom
+        if ii == 0 or (ii > 0 and config['wavelength'][ii] != config['wavelength'][ii - 1]):
             nextatom = True
-            print('Atom for wavelength')
+            atomstr += 'wavelength, '
 
-        #         if observe_class == 'science':
-
-        # AB
-        # ABBA
-        if qoffsets and nsteps >= 4 and nsteps - ii > 3 and nabba == 0:
-            #             print(q, sequence[ii+1]['telescope:q'], sequence[ii+2]['telescope:q'], sequence[ii+3]['telescope:q'])
-            if q == float(sequence[ii + 3]['telescope:q']) and q != float(sequence[ii + 1]['telescope:q']) and float(
-                    sequence[ii + 1]['telescope:q']) == float(sequence[ii + 2]['telescope:q']):
-                nabba = 3
+        # A change in exposure time or coadds is a new atom for science exposures
+        if step['observe:observeType'].upper() not in ['FLAT', 'ARC', 'DARK', 'BIAS']:
+            if observe_class.upper() == 'SCIENCE' and \
+                    ii > 0 and (exptimes[ii] != exptimes[ii_prevobj] or coadds[ii] != coadds[ii_prevobj]):
                 nextatom = True
-                print('Atom for ABBA')
-        else:
-            nabba -= 1
-
+                atomstr += 'exposure time change, '
+
+            # Offsets - a new offset pattern is a new atom
+            #         print('npattern: ', npattern)
+            if not (offset_lag == 0 and exptime_groups == True):
+                # For NIR imaging, need to have at least two offset positions if no repeating pattern
+                # New atom after every 2nd offset (noffsets is odd)
+                if mode == 'imaging' and offset_lag == 0 and all([w > 1.0 for w in config['wavelength']]):
+                    if ii == 0:
+                        noffsets += 1
+                    else:
+                        if poffsets[ii] != poffsets[ii_prevobj] or qoffsets[ii] != qoffsets[ii_prevobj]:
+                            noffsets += 1
+                    if noffsets % 2 == 1:
```

Comment:
It's not explicitly clear to me why `noffsets` being odd has the special case of starting a next atom?